### PR TITLE
lang: Fix ValueOf() with nil pointer values

### PR DESCRIPTION
**Pull request by frebib**
Some forms of reflect.Value can cause ValueOf() to panic when there is a
nil pointer somewhere within the reflect.Value, whether that be a
container type like a struct, list or map, or just a raw nil pointer.

In these cases, ValueOf() attempted to dereference the pointer without
ever checking if it was nil. mgmt lang doesn't have pointers of any
kind, so the equivalent is the zero type of the value. In such cases,
gracefully attempt to handle by constructing a types.Value based on a
new instance of the type of the pointer reflect.Value instead.

Signed-off-by: Joe Groocock <me@frebib.net>

### CODE REVIEW

**Comment by purpleidea** [comment]
Optional nit about two sentences with missing periods and one capitalization perhaps, but I can merge as is if tests are relatively happy =D

Thanks for the contribution!

**Comment by purpleidea** [comment]
(I don't know why my brain even sees these things.)

**Comment by frebib** [comment]
> Optional nit about two sentences with missing periods and one capitalization perhaps, but I can merge as is if tests are relatively happy =D
> 
> Thanks for the contribution!

Haha I knew you'd pick up on this. I even try super hard when writing comments to try and do the right thing but I think I struggle to see the distinction. Will update

**Comment by purpleidea** [comment]
I don't mind merging as-is.

File: lang/types/value.go [hunk 1]
```diff
@@ -153,6 +164,14 @@ func ValueOf(v reflect.Value) (Value, error) {
 		values := make(map[string]Value)
 		for i := 0; i < l; i++ {
 			x := value.Field(i)
+			// It's likely that setting the field is desirable, so clone the
+			// field to prevent the recursive ValueOf() call failing. This is
+			// obviously slower so hope we don't need this often.
+			if !x.CanSet() {
+				c := reflect.New(x.Type()).Elem()
```
Comment (frebib) [inline lang/types/value.go:171]: I'm still not 100% on this, but it seems to work as-is. I feel like I'm inferring `x.Kind() == Ptr` from `!x.CanSet()` which may or may not hold true in the majority of cases. I think the thing I'm not sure on is _why_ `x.CanSet() == false`. Debugging this isn't easy when it's all just pointers

**Comment by frebib** [comment]
CI seems happy now. I fixed up the comments

File: lang/types/value.go [hunk 2]
```diff
@@ -58,9 +58,20 @@ func ValueOf(v reflect.Value) (Value, error) {
 	typ := value.Type()
 	kind := typ.Kind()
 	for kind == reflect.Ptr {
+		// Not much that can be done here. Caller must pass a Settable Ptr value.
+		if !value.CanSet() {
```
Comment (purpleidea) [inline lang/types/value.go:62]: I'm not 100% sure this is what you want. In situations like these, what I'd recommend doing, is adding a single simple test with the short, standalone bit that you want to test:

Like:
TestFrebibNewTest(t....) ...  {
v, err := ValueOf(XXX)
if err != nil ...
if v == nil ...
if err := v.Cmp(XXX); err != nil ...
}

I think that would help us better understand where you're stuck, and then we can go through and understand the reflect lib and which method you want a bit better.

**Review by purpleidea — CHANGES_REQUESTED** [review]
I was bit sleepy last night so not really reading code properly. I think it's worth looking into this. I can help you solve it if you want.

I'm not convinced about `CanSet` initially, because it's not `value` that we want to set... In fact, we're building a new Value here, there's no modifying or "Set()" happening (in reflect terms) ... Of course I could be overlooking the use of CanSet to determine something else, so let's simplify this by defining exactly what's wrong with a standalone test case first.

(And I'm guessing something in your new docker resource code triggered this, so feel free to post that for a quick look in case the bug is there.)

File: lang/types/value.go (hunk 2 above)
Comment (frebib) [reply lang/types/value.go]: You're right. I was seriously overthinking this. Reworked it to a really simple change. Most of the diff is just adjusting the tests now.

**Comment by frebib** [comment]
This still doesn't fix the crashes :(
reflect is doing something weird with nil pointers here that I don't understand

**Comment by purpleidea** [comment]
Did you want a review on this?

One test is failing:

>   lang/types/value_test.go:617:26:warning: should drop = nil from declaration of var nilStrPtr; it is the zero value (golint)

**Comment by purpleidea** [comment]
I wasn't 100% sure what test was failing previously that this patch was meant to fix, so I merged a patch with two new standalone tests... You'll note they both pass. This is what I had in mind just to make it very obvious to see what's going on...

33fda8605a7e46597651edc2834c2dae618ed6ee

All the table tests and wrapper functions are great for writing lots of tests, but in my simplistic brain this makes it easier to find/work on specific issues...

In any case, it all is passing. LMK how I can help.

**Comment by frebib** [comment]
Tests should be fixed now. Assuming CI passes, this should be ready for review 🙏🏻

**Comment by frebib** [comment]
After discussion, we came to the conclusion that this is probably not the correct fix. Instead the behaviour is correct, but the panic is not prevented/handled appropriately, to be fixed in https://github.com/purpleidea/mgmt/pull/655

**Pull request closed without merging**